src/simpleShape.cc

```diff
@@ -292,61 +293,114 @@ MatrixM SimpleShape::correctWeightedMoments(
+                                                            std::vector<std::string> ({"Ixx", "Iyy", "Ixy",
+                                                                                      "Ix", "Iy"}), "pixels");
+            r._flagHandler = lsst::meas::base::FlagHandler::addFields(schema,
+                                                            name, flagDefs.begin(), flagDefs.end());
+            return r;
+}
+
+SimpleShapeResultKey::SimpleShapeResultKey(lsst::afw::table::SubSchema const & s) :
+    _shapeResult(s),
+    _centroidResult(s),
+    _uncertantyResult(s, std::vector<std::string> ({"Ixx", "Iyy", "Ixy", "Ix", "Iy"})),
+    _flagHandler(s, flagDefs.begin(), flagDefs.end())
 {}
 
-template<typename PixelT>
-void SimpleShape::_apply(
+SimpleShapeResult SimpleShapeResultKey::get(lsst::afw::table::BaseRecord const & record) const {
+    SimpleShapeResult result;
+    result.ellipse = record.get(_shapeResult);
+    result.center = record.get(_centroidResult);
+    result.covariance = record.get(_uncertantyResult);
+    for (int n = 0; n < SimpleShape::N_FLAGS; ++n) {
+        result.flags[n] = _flagHandler.getValue(record, n);
+    }
+    return result;
+}
+
+void SimpleShapeResultKey::set(afw::table::BaseRecord & record, SimpleShapeResult const & value) const {
+    record.set(_shapeResult, value.ellipse);
+    record.set(_centroidResult, value.center);
+    record.set(_uncertantyResult, value.covariance);
+    for (int n = 0; n < SimpleShape::N_FLAGS; ++n) {
+        _flagHandler.setValue(record, n, value.flags[n]);
+    }
+}
+
+bool SimpleShapeResultKey::operator==(SimpleShapeResultKey const & other) const {
+    return _shapeResult == other._shapeResult &&
+        _centroidResult == other._centroidResult &&
+        _uncertantyResult == other._uncertantyResult;
+    //don't bother with flags - if we've gotten this far, it's basically impossible the flags don't match
+}
+
+bool SimpleShapeResultKey::isValid() const {
+    return _shapeResult.isValid() &&
+        _centroidResult.isValid() &&
+        _uncertantyResult.isValid();
+    //don't bother with flags - if we've gotten this far, it's basically impossible the flags don't match
+}
+
+
+SimpleShape::SimpleShape(
+        Control const & ctrl,
+        std::string const & name,
+        afw::table::Schema & schema
+    )
+    : _ctrl(ctrl),
+      _resultKey(ResultKey::addFields(schema, name)),
+      _centroidExtractor(schema, name)
+    {}
+
+void SimpleShape::measure(
     afw::table::SourceRecord & source,
-    afw::image::Exposure<PixelT> const & exposure,
-    afw::geom::Point2D const & center
+    afw::image::Exposure<float> const & exposure
 ) const {
-    SimpleShapeControl const & ctrl = static_cast<SimpleShapeControl const &>(this->getControl());
+    afw::geom::Point2D center = _centroidExtractor(source, _resultKey.getFlagHandler());
+    afw::geom::ellipses::Ellipse weight(afw::geom::ellipses::Axes(_ctrl.sigma), center);
```

Comment:
I suggest `auto` in the two lines above. Please look for other instances where it will also work. It clarifies the code.